File: android/src/main/java/com/getchange/zendesk_flutter_plugin/ZendeskFlutterPlugin.java
package com.getchange.zendesk_flutter_plugin;

import android.os.Handler;
import android.os.Looper;
import android.text.TextUtils;

import com.google.gson.GsonBuilder;

import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.PluginRegistry;
import zendesk.chat.Account;
import zendesk.chat.Agent;
import zendesk.chat.Chat;
import zendesk.chat.ChatLog;
import zendesk.chat.ChatProvider;
import zendesk.chat.ChatRating;
import zendesk.chat.ChatState;
import zendesk.chat.ConnectionStatus;
import zendesk.chat.ObservationScope;
import zendesk.chat.Observer;
import zendesk.chat.OfflineForm;
import zendesk.chat.ProfileProvider;
import zendesk.chat.VisitorInfo;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static com.google.gson.FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES;


public class ZendeskFlutterPlugin implements MethodCallHandler {

  private Handler mainHandler = new Handler(Looper.getMainLooper());
  private PluginRegistry.Registrar registrar;
  private String applicationId = null;

  private ObservationScope connectionScope = null;
  private ObservationScope accountScope = null;
  private ObservationScope chatScope = null;

  private ZendeskFlutterPlugin.EventChannelStreamHandler connectionStreamHandler = new ZendeskFlutterPlugin.EventChannelStreamHandler();
  private ZendeskFlutterPlugin.EventChannelStreamHandler accountStreamHandler = new ZendeskFlutterPlugin.EventChannelStreamHandler();
  private ZendeskFlutterPlugin.EventChannelStreamHandler agentsStreamHandler = new ZendeskFlutterPlugin.EventChannelStreamHandler();
  private ZendeskFlutterPlugin.EventChannelStreamHandler chatItemsStreamHandler = new ZendeskFlutterPlugin.EventChannelStreamHandler();

  private static class EventChannelStreamHandler implements EventChannel.StreamHandler {
    private EventChannel.EventSink eventSink = null;

    public void success(Object event) {
      if (eventSink != null) {
        eventSink.success(event);
      }
    }

    public void error(String errorCode, String errorMessage, Object errorDetails) {
      if (eventSink != null) {
        eventSink.error(errorCode, errorMessage, errorDetails);
      }
    }

    @Override
    public void onListen(Object o, EventChannel.EventSink eventSink) {
      this.eventSink = eventSink;
    }

    @Override
    public void onCancel(Object o) {
      this.eventSink = null;
    }
  }

  public static void registerWith(PluginRegistry.Registrar registrar) {
    final MethodChannel callsChannel = new MethodChannel(registrar.messenger(), "plugins.flutter.zendesk_chat_api/calls");
    final EventChannel connectionStatusEventsChannel = new EventChannel(registrar.messenger(), "plugins.flutter.zendesk_chat_api/connection_status_events");
    final EventChannel accountStatusEventsChannel = new EventChannel(registrar.messenger(),"plugins.flutter.zendesk_chat_api/account_status_events");
    final EventChannel agentEventsChannel = new EventChannel(registrar.messenger(),"plugins.flutter.zendesk_chat_api/agent_events");
    final EventChannel chatItemsEventsChannel = new EventChannel(registrar.messenger(),"plugins.flutter.zendesk_chat_api/chat_items_events");

    ZendeskFlutterPlugin plugin = new ZendeskFlutterPlugin(registrar);

    callsChannel.setMethodCallHandler(plugin);

    connectionStatusEventsChannel.setStreamHandler(plugin.connectionStreamHandler);
    accountStatusEventsChannel.setStreamHandler(plugin.accountStreamHandler);
    agentEventsChannel.setStreamHandler(plugin.agentsStreamHandler);
    chatItemsEventsChannel.setStreamHandler(plugin.chatItemsStreamHandler);
  }

  private ZendeskFlutterPlugin(PluginRegistry.Registrar registrar) {
    this.registrar = registrar;
  }

  @Override
  public void onMethodCall(MethodCall call, MethodChannel.Result result) {
    switch(call.method) {
      case "getPlatformVersion":
        result.success("Android " + android.os.Build.VERSION.RELEASE);
        break;
      case "init":
        applicationId = call.argument("applicationId");
        final String accountKey = call.argument("accountKey");
        try {
          Chat.INSTANCE.init(registrar.activity(), accountKey);
        } catch (Exception e) {
          result.error("UNABLE_TO_INITIALIZE_CHAT_API", e.getMessage(), e);
          break;
        }
        result.success(null);
        break;
      case "startChat":
        ProfileProvider profileProvider = Chat.INSTANCE.providers().profileProvider();
        ChatProvider chatProvider = Chat.INSTANCE.providers().chatProvider();

        VisitorInfo visitorInfo = VisitorInfo.builder()
                .withPhoneNumber(call.argument("visitorPhone"))
                .withEmail(call.argument("visitorEmail"))
                .withName(call.argument("visitorName"))
                .build();

        profileProvider.setVisitorInfo(visitorInfo, null);

        String department = call.argument("department");
        String tags = call.argument("tags");

        if (!TextUtils.isEmpty(department)) {
          chatProvider.setDepartment(department, null);
        }
        if (!TextUtils.isEmpty(tags)) {
          profileProvider.addVisitorTags(Arrays.asList(tags.split(",")), null);
        }

        bindChatListeners();

        Chat.INSTANCE.providers().connectionProvider().connect();

        result.success(null);
        break;
      case "endChat":
          unbindChatListeners();
          Chat.INSTANCE.providers().chatProvider().endChat(null);
          result.success(null);
        break;
      case "sendMessage":
        if (Chat.INSTANCE.providers().connectionProvider().getConnectionStatus() != ConnectionStatus.CONNECTED) {
          result.error("CHAT_NOT_STARTED", null, null);
        } else {
          String message = call.argument("message");
          Chat.INSTANCE.providers().chatProvider().sendMessage(message);
          result.success(null);
        }
        break;
      case "resendMessage":
        if (Chat.INSTANCE.providers().connectionProvider().getConnectionStatus() != ConnectionStatus.CONNECTED) {
          result.error("CHAT_NOT_STARTED", null, null);
        } else {
          String messageId = call.argument("messageId");
          Chat.INSTANCE.providers().chatProvider().resendFailedMessage(messageId);
          result.success(null);
        }
        break;
      case "sendComment":
        if (Chat.INSTANCE.providers().connectionProvider().getConnectionStatus() != ConnectionStatus.CONNECTED) {
          result.error("CHAT_NOT_STARTED", null, null);
        } else {
          String comment = call.argument("comment");
          Chat.INSTANCE.providers().chatProvider().sendChatComment(comment, null);
          result.success(null);
        }
        break;
      case "sendAttachment":
        if (Chat.INSTANCE.providers().connectionProvider().getConnectionStatus() != ConnectionStatus.CONNECTED) {
          result.error("CHAT_NOT_STARTED", null, null);
        } else {
          String pathname = call.argument("pathname");
          if (TextUtils.isEmpty(pathname)) {
            result.error("ATTACHMENT_EMPTY_PATHNAME", null, null);
            return;
          }
          File file = new File(pathname);
          if (!file.isFile()) {
            result.error("ATTACHMENT_NOT_FILE", null, null);
            return;
          }
          Chat.INSTANCE.providers().chatProvider().sendFile(file, null);
          result.success(null);
        }
        break;
      case "sendChatRating": {
        if (Chat.INSTANCE.providers().connectionProvider().getConnectionStatus() != ConnectionStatus.CONNECTED) {
          result.error("CHAT_NOT_STARTED", null, null);
          return;
        }
        ChatRating chatLogRating = null;
        ChatProvider provider = Chat.INSTANCE.providers().chatProvider();
        String rating = call.argument("rating");
        if (!TextUtils.isEmpty(rating)) {
          chatLogRating = toChatLogRating(rating);
        }

        if (chatLogRating != null) {
          provider.sendChatRating(chatLogRating, null);
        }

        String comment = call.argument("comment");
        if (!TextUtils.isEmpty(comment)) {
          provider.sendChatComment(comment, null);
        }
        result.success(null);
        break;
      }
      case "sendOfflineMessage":
        if (Chat.INSTANCE.providers().connectionProvider().getConnectionStatus() != ConnectionStatus.CONNECTED) {
          result.error("CHAT_NOT_STARTED", null, null);
          return;
        }
        VisitorInfo info = Chat.INSTANCE.providers().profileProvider().getVisitorInfo();
        if (TextUtils.isEmpty(info.getEmail())) {
          result.error("VISITOR_EMAIL_MUST_BE PROVIDED", null, null);
          return;
        }

        Chat.INSTANCE.providers().chatProvider().sendOfflineForm(OfflineForm.builder(call.argument("message")).withVisitorInfo(info).build(), null);

        result.success(null);

        break;
      default:
        result.notImplemented();
    }
  }

  private void bindChatListeners() {
    unbindChatListeners();

    connectionScope = new ObservationScope();
    Chat.INSTANCE.providers().connectionProvider().observeConnectionStatus(connectionScope, new Observer<ConnectionStatus>() {
      @Override
      public void update(ConnectionStatus status) {
        mainHandler.post(() -> {
          connectionStreamHandler.success(status.name());
        });
      }
    });

    accountScope = new ObservationScope();
    Chat.INSTANCE.providers().accountProvider().observeAccount(accountScope, new Observer<Account>() {
      @Override
      public void update(Account account) {
        mainHandler.post(() -> {
          accountStreamHandler.success(account.getStatus().name());
        });
      }
    });

    chatScope = new ObservationScope();
    Chat.INSTANCE.providers().chatProvider().observeChatState(chatScope, new Observer<ChatState>() {
      @Override
      public void update(ChatState chatState) {
        List<ChatAgent> agents = new ArrayList<>();

        for (Agent agent: chatState.getAgents()) {
          agents.add(ChatAgent.fromAgent(agent));
        }

        mainHandler.post(() -> {
          agentsStreamHandler.success(toJson(agents));
        });

        List<ChatLogEvent> chatLogs = new ArrayList<>();

        for (ChatLog chatLog: chatState.getChatLogs()) {
          chatLogs.add(ChatLogEvent.fromChatLog(chatLog));
        }

        mainHandler.post(() -> {
          chatItemsStreamHandler.success(toJson(chatLogs));
        });
      }
    });
  }

  private void unbindChatListeners() {
    if (connectionScope != null && !connectionScope.isCancelled()) {
      connectionScope.cancel();
      connectionScope = null;
    }
    if (chatScope != null && !chatScope.isCancelled()) {
      chatScope.cancel();
      chatScope = null;
    }
    if (accountScope != null && !accountScope.isCancelled()) {
      accountScope.cancel();
      accountScope = null;
    }
  }

  private String toJson(Object object) {
    return new GsonBuilder()
        .setFieldNamingPolicy(LOWER_CASE_WITH_UNDERSCORES)
        .create()
        .toJson(object)
        .replaceAll("\\$(string|int|bool)\":", "\":");
  }

  private ChatRating toChatLogRating(String rating) {
    switch (rating) {
      case "ChatRating.GOOD":
        return ChatRating.GOOD;
      case "ChatRating.BAD":
        return ChatRating.BAD;
      default:
        return null;
    }
  }
}

File: ios/Classes/ZendeskFlutterPlugin.m
#import "ZendeskFlutterPlugin.h"

@implementation EventChannelStreamHandler

- (void) send:(NSObject*)event {
  if (self.eventSink) {
    self.eventSink(event);
  }
}

- (FlutterError* _Nullable)onListenWithArguments:(id _Nullable)arguments
                                       eventSink:(FlutterEventSink)events {
  self.eventSink = events;
  return nil;
}

- (FlutterError* _Nullable)onCancelWithArguments:(id _Nullable)arguments {
  _eventSink = nil;
  return nil;
}

@end

@implementation ZendeskFlutterPlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* callsChannel = [FlutterMethodChannel
      methodChannelWithName:@"plugins.flutter.zendesk_chat_api/calls"
      binaryMessenger:[registrar messenger]];

  FlutterEventChannel* connectionStatusEventsChannel = [FlutterEventChannel
      eventChannelWithName:@"plugins.flutter.zendesk_chat_api/connection_status_events"
      binaryMessenger:[registrar messenger]];

  FlutterEventChannel* accountStatusEventsChannel = [FlutterEventChannel
      eventChannelWithName:@"plugins.flutter.zendesk_chat_api/account_status_events"
      binaryMessenger:[registrar messenger]];

  FlutterEventChannel* agentEventsChannel = [FlutterEventChannel
      eventChannelWithName:@"plugins.flutter.zendesk_chat_api/agent_events"
      binaryMessenger:[registrar messenger]];
      
  FlutterEventChannel* chatItemsEventsChannel = [FlutterEventChannel
      eventChannelWithName:@"plugins.flutter.zendesk_chat_api/chat_items_events"
      binaryMessenger:[registrar messenger]];

  ZendeskFlutterPlugin* instance = [[ZendeskFlutterPlugin alloc] init];
  
  instance.connectionStreamHandler = [[EventChannelStreamHandler alloc] init];
  instance.accountStreamHandler = [[EventChannelStreamHandler alloc] init];
  instance.agentsStreamHandler = [[EventChannelStreamHandler alloc] init];
  instance.chatItemsStreamHandler = [[EventChannelStreamHandler alloc] init];
  
  [registrar addMethodCallDelegate:instance channel:callsChannel];
  [connectionStatusEventsChannel setStreamHandler:instance.connectionStreamHandler];
  [accountStatusEventsChannel setStreamHandler:instance.accountStreamHandler];
  [agentEventsChannel setStreamHandler:instance.agentsStreamHandler];
  [chatItemsEventsChannel setStreamHandler:instance.chatItemsStreamHandler];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"getPlatformVersion" isEqualToString:call.method]) {
    result([@"iOS " stringByAppendingString:[[UIDevice currentDevice] systemVersion]]);
  } else if ([@"init" isEqualToString:call.method]) {
    self.accountKey = call.arguments[@"accountKey"];
    result(nil);
  } else if ([@"startChat" isEqualToString:call.method]) {
    if ([self.accountKey length] == 0) {
      result([FlutterError errorWithCode:@"NOT_INITIALIZED" message:nil details:nil]);
      return;
    }
    
    if (ZDKChat.instance) {
      result([FlutterError errorWithCode:@"CHAT_SESSION_ALREADY_OPEN" message:nil details:nil]);
      return;
    }
      
    [ZDKChat initializeWithAccountKey:self.accountKey queue:dispatch_get_main_queue()];
      
    ZDKChatAPIConfiguration *chatAPIConfiguration = [[ZDKChatAPIConfiguration alloc] init];
    chatAPIConfiguration.department = [self argumentAsString:call forName:@"department"];
    chatAPIConfiguration.visitorInfo = [[ZDKVisitorInfo alloc] initWithName:[self argumentAsString:call forName:@"visitorName"]
                                                                        email:[self argumentAsString:call forName:@"visitorEmail"]
                                                                  phoneNumber:[self argumentAsString:call forName:@"visitorPhone"]];
    
    NSString *tags = [self argumentAsString:call forName:@"tags"];
    if ([tags length] != 0) {
      chatAPIConfiguration.tags = [tags componentsSeparatedByString:@","];
    }

    ZDKChat.instance.configuration = chatAPIConfiguration;
    
    [self bindChatListeners];
    [ZDKChat.connectionProvider connect];
    result(nil);
  } else if ([@"endChat" isEqualToString:call.method]) {
    if (ZDKChat.instance != nil) {
      [self unbindChatListeners];
      [ZDKChat.chatProvider endChat:nil];
    }
    result(nil);
  } else if ([@"sendMessage" isEqualToString:call.method]) {
    if (ZDKChat.instance == nil) {
      result([FlutterError errorWithCode:@"CHAT_NOT_STARTED" message:nil details:nil]);
      return;
    }
    [ZDKChat.chatProvider sendMessage:call.arguments[@"message"] completion:nil];
    result(nil);
  } else if ([@"resendMessage" isEqualToString:call.method]) {
     if (ZDKChat.instance == nil) {
       result([FlutterError errorWithCode:@"CHAT_NOT_STARTED" message:nil details:nil]);
       return;
     }
      //[ZDKChat.chatProvider resendFailedMessageWithId:call.arguments[@"messageId"]] completion:nil];
      result(nil);
   } else if ([@"sendComment" isEqualToString:call.method]) {
      if (ZDKChat.instance == nil) {
        result([FlutterError errorWithCode:@"CHAT_NOT_STARTED" message:nil details:nil]);
        return;
      }
      [ZDKChat.chatProvider sendChatComment:call.arguments[@"comment"] completion:nil];
      result(nil);
    } else if ([@"sendAttachment" isEqualToString:call.method]) {
    if (ZDKChat.instance == nil) {
      result([FlutterError errorWithCode:@"CHAT_NOT_STARTED" message:nil details:nil]);
      return;
    }
    NSString* pathname = [self argumentAsString:call forName:@"pathname"];
    if ([pathname length] == 0) {
      result([FlutterError errorWithCode:@"ATTACHMENT_EMPTY_PATHNAME" message:nil details:nil]);
      return;
    }
    NSString* filename = [pathname lastPathComponent];
    NSFileManager* filemgr = [NSFileManager defaultManager];
    if (![filemgr fileExistsAtPath:pathname]) {
      result([FlutterError errorWithCode:@"ATTACHMENT_FILE_MISSING" message:nil details:nil]);
      return;
    }
    [ZDKChat.chatProvider sendFileWithUrl:[filemgr contentsAtPath:pathname] onProgress:nil completion:nil];
    result(nil);
  } else if ([@"sendChatRating" isEqualToString:call.method]) {
    if (ZDKChat.instance == nil) {
      result([FlutterError errorWithCode:@"CHAT_NOT_STARTED" message:nil details:nil]);
      return;
    }
    NSString* rating = [self argumentAsString:call forName:@"rating"];
    [ZDKChat.chatProvider sendChatRating:[self toChatLogRating:rating] completion:nil];
    NSString* comment = [self argumentAsString:call forName:@"comment"];
    if (comment != nil) {
        [ZDKChat.chatProvider sendChatComment:comment completion:nil];
    }
    result(nil);
  } else if ([@"sendOfflineMessage" isEqualToString:call.method]) {
    if (ZDKChat.instance == nil) {
      result([FlutterError errorWithCode:@"CHAT_NOT_STARTED" message:nil details:nil]);
      return;
    }
    ZDKOfflineForm *form = [[ZDKOfflineForm alloc] initWithVisitorInfo:ZDKChat.instance.profileProvider.visitorInfo
                                                        departmentId:ZDKChat.instance.configuration.department
                                                        message:call.arguments[@"message"]];
    [ZDKChat.chatProvider sendOfflineForm:form completion:nil];
    result(nil);
  } else {
    result(FlutterMethodNotImplemented);
  }
}

- (NSString*) argumentAsString:(FlutterMethodCall*)call forName:(NSString*)argName {
  if ([call.arguments isKindOfClass:[NSNull class]]) {
    return nil;
  }
  NSString* value = call.arguments[argName];
  return [value isKindOfClass:[NSString class]] ? value : nil;
}

- (NSData*) argumentAsBinary:(FlutterMethodCall*)call forName:(NSString*)argName {
  if ([call.arguments isKindOfClass:[NSNull class]]) {
    return nil;
  }
  NSObject* value = call.arguments[argName];
  if (![value isKindOfClass:[FlutterStandardTypedData class]]) {
    return nil;
  }
  return ((FlutterStandardTypedData*)value).data;
}

- (void) bindChatListeners {
    [self unbindChatListeners];
    
    _accountToken = [ZDKChat.accountProvider observeAccount:^(ZDKChatAccount *account) {
        [self.accountStreamHandler send:(account.accountStatus == ZDKChatAccountStatusOnline ? @("ONLINE") : @("OFFLINE"))];
    }];
    
    _chatToken = [ZDKChat.chatProvider observeChatState:^(ZDKChatState *chatState) {
        NSMutableArray* outAgents = [[NSMutableArray alloc] initWithCapacity:[chatState.agents count]];
        NSLog(@"Agent count:%tu", chatState.agents.count);
        [chatState.agents enumerateObjectsUsingBlock:^(ZDKAgent *agent, NSUInteger index, BOOL *stop) {
          NSMutableDictionary *agentDict = [[NSMutableDictionary alloc] init];
          [agentDict setValue:agent.displayName forKey:@"display_name"];
          [agentDict setValue:agent.avatar forKey:@"avatar_path"];
          [agentDict setValue:@(agent.isTyping) forKey:@"is_typing"];
          [agentDict setValue:agent.nick forKey:@"nick"];
          [outAgents addObject:agentDict];
        }];
        [self.agentsStreamHandler send:[self toJson:outAgents]];
        
        NSMutableArray* outLogs = [[NSMutableArray alloc] initWithCapacity:[chatState.logs count]];
        NSLog(@"Log count:%tu", chatState.logs.count);
        for (ZDKChatLog* event in chatState.logs) {
          NSMutableDictionary* chatItem = [[NSMutableDictionary alloc] init];
          [chatItem setValue:event.id forKey:@"id"];
          [chatItem setValue:@(event.createdTimestamp) forKey:@"create_timestamp"];
          [chatItem setValue:@(event.lastModifiedTimestamp) forKey:@"modify_timestamp"];
          [chatItem setValue:event.nick forKey:@"nick"];
          [chatItem setValue:event.displayName forKey:@"display_name"];
          [chatItem setValue:[self chatEventTypeToString:event.type] forKey:@"type"];
          [chatItem setValue:[self deliveryStatusToString:event.deliveryStatus] forKey:@"delivery_status"];
          [chatItem setValue:[self participantToString:event.participant] forKey:@"participant"];
            
          if ([chatItem isMemberOfClass:[ZDKChatMessage class]]) {
                ZDKChatMessage *message = (ZDKChatMessage *)chatItem;
                [chatItem setValue:message.message forKey:@"message"];
          }
            
          if ([chatItem isMemberOfClass:[ZDKChatRating class]]) {
                ZDKChatRating *rating = (ZDKChatRating *)chatItem;
                [chatItem setValue:[self chatLogRatingToString:rating.ratingValue] forKey:@"current_rating"];
          }
            
          if ([chatItem isMemberOfClass:[ZDKChatAttachmentMessage class]]) {
                //[chatItem setValue:[self attachmentToDictionary:event.attachment] forKey:@"attachment"];
          }
            
          if ([chatItem isMemberOfClass:[ZDKChatComment class]]) {
                ZDKChatComment *comment = (ZDKChatComment *)chatItem;
                [chatItem setValue:comment.comment forKey:@"new_comment"];
          }
          
          [outLogs addObject:chatItem];
        }
        [self.chatItemsStreamHandler send:[self toJson:outLogs]];
    }];
    
    _chatToken = [ZDKChat.connectionProvider observeConnectionStatus:^(enum ZDKConnectionStatus status) {
        NSString *value;
        switch (status) {
          case ZDKConnectionStatusConnecting:
            value = @("CONNECTING");
            break;
          case ZDKConnectionStatusConnected:
            value = @("CONNECTED");
            break;
          case ZDKConnectionStatusFailed:
            value = @("FAILED");
            break;
          case ZDKConnectionStatusDisconnected:
            value = @("DISCONNECTED");
            break;
          case ZDKConnectionStatusUnreachable:
            value = @("UNREACHABLE");
            break;
          case ZDKConnectionStatusReconnecting:
            value = @("RECONNECTING");
            break;
          default:
            value = @("UNKNOWN");
            break;
        }
        [self.connectionStreamHandler send:value];
    }];
}

- (void) unbindChatListeners {
  [_connectionToken cancel];
  [_accountToken cancel];
  [_chatToken cancel];
}

- (NSString*) chatEventTypeToString:(ZDKChatLogType)type {
  switch (type) {
    case ZDKChatLogTypeMessage:
      return @"MESSAGE";
    case ZDKChatLogTypeAttachmentMessage:
      return @"ATTACHMENT_MESSAGE";
    case ZDKChatLogTypeMemberJoin:
      return @"MEMEBER_JOIN";
    case ZDKChatLogTypeMemberLeave:
      return @"MEMBER_LEAVE";
    case ZDKChatLogTypeChatComment:
      return @"COMMENT";
    case ZDKChatLogTypeChatRating:
      return @"RATING";
    case ZDKChatLogTypeChatRatingRequest:
      return @"RATING_REQUEST";
    default:
      return @"UNKNOWN";
  }
}

- (NSString*) deliveryStatusToString:(ZDKDeliveryStatus)status {
  switch (status) {
    case ZDKDeliveryStatusPending:
      return @"PENDING";
    case ZDKDeliveryStatusDelivered:
      return @"DELIVERED";
    case ZDKDeliveryStatusFailed:
      return @"FAILED";
    default:
      return @"UNKNOWN";
  }
}

- (NSString*) participantToString:(ZDKChatParticipant)participant {
  switch (participant) {
    case ZDKChatParticipantVisitor:
      return @"VISITOR";
    case ZDKChatParticipantAgent:
      return @"AGENT";
    case ZDKChatParticipantTrigger:
      return @"TRIGGER";
    case ZDKChatParticipantSystem:
      return @"SYSTEM";
    default:
      return @"UNKNOWN";
  }
}

/*- (NSDictionary*) attachmentToDictionary:(ZDKChatAttachment*)attachment {
  if (attachment == nil) {
    return nil;
  }
  NSMutableDictionary* out = [[NSMutableDictionary alloc] init];
  [out setValue:attachment.url forKey:@"url"];
  [out setValue:attachment.thumbnailURL forKey:@"thumbnail_url"];
  [out setValue:attachment.fileSize forKey:@"size"];
  [out setValue:attachment.mimeType forKey:@"mime_type"];
  [out setValue:attachment.fileName forKey:@"name"];
  return out;
}*/

- (NSString*) toJson:(NSObject*)object {
  NSError *error = nil;
  NSData *jsonData = [NSJSONSerialization dataWithJSONObject:object options:0 error:&error];
  if (error != nil) {
    NSLog(@("An json serialization error happened: %@"), error);
    return nil;
  } else if ([jsonData length] == 0) {
    return nil;
  } else {
    return [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
  }
}

- (ZDKRating) toChatLogRating:(NSString*) rating {
  if ([@"ChatRating.GOOD" isEqualToString:rating]) {
    return ZDKRatingGood;
  } else if ([@"ChatRating.BAD" isEqualToString:rating]) {
    return ZDKRatingBad;
  } else {
    return ZDKRatingNone;
  }
}

- (NSString*) chatLogRatingToString:(ZDKRating)rating {
  switch (rating) {
    case ZDKRatingGood:
      return @"GOOD";
    case ZDKRatingBad:
      return @"BAD";
    default:
      return nil;
  }
}

@end
